torch/utils/checkpoint.py

```diff
@@ -219,7 +219,8 @@ def checkpoint(function, *args, use_reentrant: bool = True, **kwargs):
             If ``use_reentrant=False`` is specified, ``checkpoint`` will use an
             implementation that does not require re-entrant autograd. This
             allows ``checkpoint`` to support additional functionality, such as
-            working as expected with ``torch.autograd.grad``. Note that future
+            working as expected with ``torch.autograd.grad`` and support for
+            keyword arguments input into the checkpointed function. Note that future
```

Comment:
I think you should either explicitly document that `preserve_rng_state` and `use_reentrant` must NOT be used as a kwarg by the user module or we should pass the kwargs as a kwarg argument.

I think the second is better because otherwise, we will never be able to add any new kwarg to this function in a non-BC way...

Reply:
I feel like the kwarg names we add are not very likely to conflict with kwargs in the user's module as they will likely be activation checkpoint specific? The current way seems the simplest and in-line with how we pass non-keyword args.

Although if we feel strongly about this, we can definitely switch it over to a specific `kwarg` argument.

Reply:
The problem is that "not very likely" to be used means that it is very likely to happen at least once in practice haha
And this will lead to a very surprising error for the user which will make it very hard and confusing for them to debug (why my kwarg disapeared??)

Reply:
The issue is that we're actually wrapping this function with the `CheckpointWrapper`, so the user does not have to change their forward pass, just need to wrap their module with `CheckpointWrapper`. Requiring users to modify their forward pass / change how arg and kwargs are passed in will be a significant pain point in applying activation checkpointing. 

To mitigate this, what about adding an argument `allow_kwargs` to this function which if set to True, we pass in kwargs as this PR does, and document the downsides that you've mentioned? Then, `CheckpointWrapper` can set this to be True and other use cases won't run into the issue you mentioned above.

Reply:
The `CheckpointWrapper` can take everything passed to it as `**kwargs` and then pass it to this function as `kwargs` right? That shouldn't be an issue.

Reply:
that makes sense, though I guess it limits the functionality to only CheckpointWrapper wrapped modules, instead of functional checkpoint.

Reply:
What do you mean? Both this function version and the wrapper version will be able to properly propagate the kwargs even if they're taken in as a dictionary.